### PR TITLE
Graphql attach files to sample mutation

### LINKED ISSUE #391: GraphQL: attachFilesToSample mutation
## Summary

As a User I would like to be able to attach files to my Samples through the API.

Depends on #390 

## Acceptance Criteria

* create a attachFilesToSample mutation which accepts the following params
  * `sampleId` which is the global id of a sample e.g. `gid://irida/Sample/12`
    * type is `ID!`
  * `files` which is an array of `signedBlobId` returned from `createDirectUpload` mutation defined in #390 
    * type is `[String!]!`
* Return the following fields
  * `errors` which is an array of `String` describing any errors encountered
  * `sample` the sample which now has the files attached to it
* The mutation should use the existing `Attachments::CreateService`

### PR DESCRIPTION
**Pull request by JeffreyThiessen**
## What does this PR do and why?

Adds AttachFilesToSample mutation to the GraphQL API
Fixes #391 

## Screenshots or screen recordings

N/A

## How to set up and validate locally

1. `./scripts/clean_dev.sh`
2. Login as `user1@email.com`
3. Navigate to `http://localhost:3000/graphiql` to open the graphql interface .
4. Verify you can query sample from graphql and rails console

This graphql query gets you sample 51
```graphql
{
query getonesamp {
  node(id: "gid://irida/Sample/51"){
    ... on Sample{
      id,
      name,
      description,
      metadata
    }
  }
}
}
```

In a terminal, open rails console with `rails c`

The following command gets you the sample object matching above
```ruby
> Sample.all[59]
```
querying for attachments should result in an empty list
```ruby
> Sample.all[59].attachments
```

5. find a file to upload, get the content type, md5 checksum and bytesize. Or download this file and use the provided values in step 6

[myfile.json](https://github.com/phac-nml/irida-next/files/14129150/myfile.json)

6. Create a blob for your file

```graphql
mutation makeblob {
  createDirectUpload(
    input: {
      filename: "myfile.json", # file name
      contentType: "application/json", # file content type
      checksum: "Y33CgI35hFoI6p+WBXYl+A==", # md5
      byteSize: 151 # size in bytes
    }
  ) {
    directUpload {
      url,
      headers,
      blobId,
      signedBlobId
    }
  }
}
```

We will use the `url`, `headers` and `signedBlobId`

7. Upload your file. This is simplest with Postman
Make a `PUT` request, putting the returned `url` in the URL field

image

Attach the file as binary

image

Verify your content type header matches the `headers` from above. Postman should auto-generate this based on the file you attached.

When `Send` is clicked, you should receive a `204 No Content`

8. Execute the mutation to attach files to a Sample.

```graphql
mutation attachfilestosample {
  attachFilesToSample (
    input: {
      files: ["YOUR signedBlobId FROM ABOVE"],
      sampleId: "gid://irida/Sample/51"
    }
  ) {
    sample{id},
    status,
    errors
  }
}
```
response
```json
{
  "data": {
    "attachFilesToSample": {
      "sample": {
        "id": "gid://irida/Sample/51"
      },
      "status": {
        "eyJfcmFpbHMiOnsiZGF0YSI6NTIsInB1ciI6ImJsb2JfaWQifX0=--1aae6b4a9798a4593e24499df749f20ad862af00": "success"
      },
      "errors": []
    }
  }
}
```

9. Verify that you see all the files you attached

Querying for attachments should result in one attachment, which you uploaded.
```ruby
> Sample.all[59].attachments
```

```ruby
[#<Attachment:0x00007fe0db6ae9d8
  id: 51,
  metadata: {"format"=>"unknown", "compression"=>"none"},
  deleted_at: nil,
  attachable_type: "Sample",
  attachable_id: 51,
  created_at: Thu, 01 Feb 2024 17:59:39.890147000 UTC +00:00,
  updated_at: Thu, 01 Feb 2024 17:59:39.921974000 UTC +00:00,
  log_data: nil>]
```

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: app/graphql/mutations/attach_files_to_sample.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    field :errors, [String], null: false, description: 'A list of errors that prevented the mutation.'
+    field :sample, Types::SampleType, null: false, description: 'The updated sample.'
+    field :status, GraphQL::Types::JSON, null: false, description: 'The status of the mutation.'
+
+    def resolve(sample_id:, files:)
+      sample = IridaSchema.object_from_id(sample_id, { expected_type: Sample })
+      files_attached = Attachments::CreateService.new(current_user, sample, { files: }).execute
+      {
+        sample:,
+        status: files_attached,
```
Comment (ericenns) [inline app/graphql/mutations/attach_files_to_sample.rb:21]: **nitpick (blocking):** Can you change `status` to return a custom json object instead of the attachments. Since if you try and upload an attachment again to a sample you get back a weird response in status which doesn't quite make sense.

e.g.

```
{
  "data": {
    "attachFilesToSample": {
      "sample": {
        "id": "gid://irida/Sample/51"
      },
      "status": [
        {
          "id": null,
          "metadata": {},
          "deleted_at": null,
          "attachable_type": "Sample",
          "attachable_id": 51,
          "created_at": null,
          "updated_at": null,
          "log_data": null
        }
      ],
      "errors": []
    }
  }
}
```

and also `errors` is not populated.

If you look at `app/controllers/projects/samples/attachments_controller.rb` if sets a status and then in the view `app/views/projects/samples/attachments/create.turbo_stream.rb` it extracts the errors.

Similar logic should be done here to build up the status, and populate the errors.

I am not quite sure what `status` should look like, I had thoughts of keying it by `signed_blob_id` but that would take a bit of inspection into the `attachment` and not sure if one to one. In general I think `status` should have a code, 3 possible and potentially include the `attachments` created, though I think that should wait for `attachment` type. Maybe report what `signed_blob_ids` were successfully attached.

File: test/graphql/attach_files_to_sample_test.rb
Comment (ericenns) [inline test/graphql/attach_files_to_sample_test.rb:1]: **nitpick (blocking):** Can you add a test where a user tries to attach the same signed_blob_id twice to a sample. I.e. tries to attach a file that is already attached. And another test where the user tries to attach a file where a file with the same checksum already exists in the Sample.

**Review by ericenns — CHANGES_REQUESTED** [review]
This is looking good, just a few comments on edge cases and error handling.

File: test/graphql/attach_files_to_sample_test.rb
Comment (JeffreyThiessen) [reply test/graphql/attach_files_to_sample_test.rb:1]: Added.
I used 2 actually different files with the same md5 for the md5 test. But I think we are safe to assume that no sequence data will ever have an md5 clash.

File: app/graphql/mutations/attach_files_to_sample.rb (hunk 1 above)
Comment (JeffreyThiessen) [reply app/graphql/mutations/attach_files_to_sample.rb]: Changes have been made.
I ended up using the signed_blob_id as the key for status value, and then made `:success` and `:error` as the status values. If you want me to change this let me know.

There is also a quirk with the `errors` returned value which MUST be an array, so right now errors output as 
```
[
  'signed--blob_id["error description"]`
]
```
Which is kind of ugly. Not sure what the best solution to pretty it would be.

Comment (ericenns) [reply app/graphql/mutations/attach_files_to_sample.rb]: Let's change the `errors` to be a `JsonType` so that we can have readable error messages.

File: test/graphql/attach_files_to_sample_test.rb
Comment (ericenns) [reply test/graphql/attach_files_to_sample_test.rb:1]: Can you also add a test for attaching a signed blob id which is invalid. Also a test for attaching files to a sample with an invalid id.

File: app/graphql/mutations/attach_files_to_sample.rb (hunk 1 above)
Comment (JeffreyThiessen) [reply app/graphql/mutations/attach_files_to_sample.rb]: Ah that should have been obvious to fix that lol. Changed now.

File: app/graphql/schema.graphql [hunk 2]
```diff
@@ -1,3 +1,48 @@
...
+type AttachFilesToSamplePayload {
+  """
+  A unique identifier for the client performing the mutation.
+  """
+  clientMutationId: String
+
+  """
+  A list of errors that prevented the mutation.
+  """
+  errors: [String!]!
```
Comment (deepsidhu85) [inline app/graphql/schema.graphql:33]: Should this not be `JSON!` since the errors is a `JSON` type?

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Just a quick comment to start that stuck out

File: app/graphql/schema.graphql (hunk 2 above)
Comment (ericenns) [reply app/graphql/schema.graphql:33]: Yeah he just forgot to run the rake task to regenerate the graphql schema after updating it which is also why the tests are failing.

File: test/graphql/attach_files_to_sample_test.rb [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    expected_status = {
+      blob_file_a.signed_id => :success,
+      blob_file_b.signed_id => :success
+    }
+    assert_equal expected_status, data['status']
+    assert_not_empty data['sample']
+
+    assert_equal 2, sample.attachments.count
+
+    # check that filenames matche
```
Comment (deepsidhu85) [inline test/graphql/attach_files_to_sample_test.rb:175]: Typo `matche`

File: test/graphql/attach_files_to_sample_test.rb [hunk 4]
```diff
@@ -0,0 +1,210 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+# good case
```
Comment (deepsidhu85) [inline test/graphql/attach_files_to_sample_test.rb:5]: You can remove these 3 lines. The test descriptions are good enough

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
A few more comments. Just need to test it out

File: app/graphql/schema.graphql (hunk 2 above)
Comment (deepsidhu85) [reply app/graphql/schema.graphql:33]: Ah makes sense!

File: test/graphql/attach_files_to_sample_test.rb (hunk 4 above)
Comment (JeffreyThiessen) [reply test/graphql/attach_files_to_sample_test.rb]: Ah whoops that was just part of my pseudo for planning the test cases, missed those 3 lines in cleanup. Fixed now.

File: test/graphql/attach_files_to_sample_test.rb (hunk 3 above)
Comment (JeffreyThiessen) [reply test/graphql/attach_files_to_sample_test.rb]: Fixed.

File: app/graphql/schema.graphql (hunk 2 above)
Comment (JeffreyThiessen) [reply app/graphql/schema.graphql]: the schema has been updated

**Review by ericenns — APPROVED** [review]
This looks good to me!

**Review by deepsidhu85 — APPROVED** [review]

**Review by deepsidhu85 — APPROVED** [review]
Looks good to me!
